stubs/python-decouple/decouple.pyi

```diff
@@ -0,0 +1,168 @@
+FALSE_VALUES: set[str]  # undocumented
+
+def read_config(parser: ConfigParser, file: TextIO) -> None: ...  # undocumented
+def strtobool(value: str | bool) -> bool: ...  # undocumented
+
+class UndefinedValueError(Exception): ...
+class Undefined: ...
+
+undefined: Undefined
+
+class Config:
+    repository: _Repository
+    def __init__(self, repository: _Repository) -> None: ...
+    @overload
+    def get(self, option: str, default: str = ...) -> str: ...
+    @overload
+    def get(self, option: str, default: _T) -> str | _T: ...
+    @overload
+    def get(self, option: str, *, cast: Callable[[str], _T]) -> _T: ...
+    @overload
+    def get(self, option: str, default: str | _T, cast: Callable[[str], _T]) -> _T: ...
+    @overload
+    def __call__(self, option: str, default: str = ...) -> str: ...
+    @overload
+    def __call__(self, option: str, default: _T) -> str | _T: ...
+    @overload
+    def __call__(self, option: str, *, cast: Callable[[str], _T]) -> _T: ...
+    @overload
+    def __call__(self, option: str, default: str | _T, cast: Callable[[str], _T]) -> _T: ...
+
+class _Repository(Protocol):  # undocumented
+    def __contains__(self, key: str) -> bool: ...
+    def __getitem__(self, key: str) -> str: ...
+
+class RepositoryEmpty:  # undocumented
+    def __init__(self, source: str = ..., encoding: str = ...) -> None: ...
+    def __contains__(self, key: str) -> bool: ...
+    def __getitem__(self, key: str) -> str: ...
+
+class RepositoryIni(RepositoryEmpty):
+    SECTION: str
+    parser: ConfigParser
+    def __init__(self, source: str, encoding: str = ...) -> None: ...
+    def __contains__(self, key: str) -> bool: ...
+    def __getitem__(self, key: str) -> str: ...
+
+class RepositoryEnv(RepositoryEmpty):
+    data: dict[str, str]
+    def __init__(self, source: str, encoding: str = ...) -> None: ...
+    def __contains__(self, key: str) -> bool: ...
+    def __getitem__(self, key: str) -> str: ...
+
+class RepositorySecret(RepositoryEmpty):
+    data: dict[str, str]
+    def __init__(self, source: str = ...) -> None: ...
+    def __contains__(self, key: str) -> bool: ...
+    def __getitem__(self, key: str) -> str: ...
+
+class AutoConfig:
+    SUPPORTED: OrderedDict[str, RepositoryEmpty]
+    encoding: str
+    search_path: str | None
+    config: Config | None
+    def __init__(self, search_path: str | None = None) -> None: ...
+    @overload
+    def __call__(self, option: str, default: str = ...) -> str: ...
+    @overload
+    def __call__(self, option: str, default: _T) -> str | _T: ...
+    @overload
+    def __call__(self, option: str, *, cast: Callable[[str], _T]) -> _T: ...
+    @overload
+    def __call__(self, option: str, default: str | _T, cast: Callable[[str], _T]) -> _T: ...
+
+config: AutoConfig
+
+class Csv(Generic[_T, _TCsv]):
+    cast: Callable[[str], _T]
+    delimiter: str
+    strip: str
+    post_process: Callable[[Iterable[Any]], _TCsv]
+    def __call__(self, value: str) -> _TCsv: ...
```

Comment:
The implementation also supports passing None. In that case, `post_process` is called with no arguments. Typing that accurately would probably require making a callback protocol for post_process.

Reply:
Good catch, I glossed over that part in my review. I am however a little concerned how well a callback protocol will behave, considering mypy would match `Csv(post_process=set)` to `Csv[Any, builtins.list[Any]]` when I tried to replace it with a callback protocol. Changing the order of overloads would so the first one is fully generic would probably solve that, but you would still end up with unsolved type vars. Although I am not sure I have tried a callback protocol with `Any` for the first argument, instead of `_T`, so maybe that will behave a bit better.